File: v4/rust/iprange-livedb/src/publication/main_file.rs
//! Atomic main-name publication and exact reservation retirement.

use crate::error;

#[cfg(unix)]
use super::namespace::regular_link_count;
use super::namespace::{sync_file, NamespaceError};
use super::output::{self, PreparedOutput};
use super::problem::Problem;
use super::reservation::{Header, Policy};
use super::reservation_file::{self, ArmedReservation};
use super::{Housekeeping, HousekeepingArtifact};

#[derive(Debug)]
pub(crate) enum Error {
    Namespace(NamespaceError),
    Sdk(error::Error),
    Output(output::Error),
    Reservation(reservation_file::Error),
    PreviousLinkCount(u64),
    ReservationLinkCount(u64),
    Gc(Problem),
    Injected,
}

#[derive(Debug)]
pub(crate) struct Failure<T> {
    pub(crate) owner: T,
    pub(crate) cause: Error,
}

#[derive(Debug)]
pub(crate) struct MainAttempt {
    pub(crate) output: PreparedOutput,
    pub(crate) reservation: ArmedReservation,
    pub(crate) main_call_started: bool,
    pub(crate) rename_succeeded: bool,
    pub(crate) desired_proven: bool,
}

#[derive(Debug)]
pub(crate) struct PublishedMain {
    pub(crate) output: PreparedOutput,
    pub(crate) reservation: ArmedReservation,
}

#[derive(Debug)]
pub(crate) struct RetiringMain {
    pub(crate) published: PublishedMain,
    pub(crate) previous_unlinked: bool,
    pub(crate) previous_retired_proven: bool,
    pub(crate) reservation_unlinked: bool,
    pub(crate) directory_synced: bool,
    pub(crate) reservation_retired_proven: bool,
    pub(crate) housekeeping: Housekeeping,
    pub(crate) visible_housekeeping: Vec<HousekeepingArtifact>,
}

#[derive(Debug)]
pub(crate) struct PublishedOutput {
    pub(crate) output: PreparedOutput,
    pub(crate) reservation_identity: super::namespace::Identity,
    pub(crate) reservation_header: Header,
    pub(crate) housekeeping: Housekeeping,
    pub(crate) visible_housekeeping: Vec<HousekeepingArtifact>,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
enum Point {
    MainRenamed,
    MainSynced,
    DirectorySynced,
    DesiredProven,
    PreviousUnlinked,
    ReservationUnlinked,
    RetirementSynced,
}

// The fixed owner carries both locked inodes without a failure-path allocation.
#[allow(clippy::result_large_err)]
pub(crate) fn publish(
    output: PreparedOutput,
    reservation: ArmedReservation,
) -> Result<PublishedMain, Failure<MainAttempt>> {
    publish_with(output, reservation, |_| Ok(()))
}

#[allow(clippy::result_large_err)]
fn publish_with(
    output: PreparedOutput,
    reservation: ArmedReservation,
    mut checkpoint: impl FnMut(Point) -> Result<(), Error>,
) -> Result<PublishedMain, Failure<MainAttempt>> {
    let mut owner = MainAttempt {
        output,
        reservation,
        main_call_started: false,
        rename_succeeded: false,
        desired_proven: false,
    };
    match publish_steps(&mut owner, &mut checkpoint) {
        Ok(()) => Ok(PublishedMain {
            output: owner.output,
            reservation: owner.reservation,
        }),
        Err(cause) => Err(Failure { owner, cause }),
    }
}

fn publish_steps(
    owner: &mut MainAttempt,
    checkpoint: &mut impl FnMut(Point) -> Result<(), Error>,
) -> Result<(), Error> {
    verify_before_main(owner)?;
    rename_main(owner)?;
    checkpoint(Point::MainRenamed)?;
    synchronize_main(owner, checkpoint)?;
    prove_main(owner, checkpoint)
}

fn verify_before_main(owner: &MainAttempt) -> Result<(), Error> {
    owner
        .reservation
        .verify_before_main(&owner.output)
        .map_err(Error::Reservation)?;
    if owner.output.previous.is_some() {
        owner
            .output
            .verify_destination_before_main()
            .map_err(Error::Output)
    } else {
        let destination = owner.output.attempt.destination();
        destination
            .directory()
            .require_absent(destination.main())
            .map_err(Error::Namespace)
    }
}

fn rename_main(owner: &mut MainAttempt) -> Result<(), Error> {
    let destination = owner.output.attempt.destination();
    owner.main_call_started = true;
    match owner.output.policy {
        Policy::FailIfExists => destination.directory().rename_noreplace(
            owner.output.attempt.name(),
            &owner.output.file,
            destination.main(),
        )?,
        Policy::ReplaceExisting => destination.directory().exchange(
            owner.output.attempt.name(),
            &owner.output.file,
            destination.main(),
        )?,
        Policy::ReplaceExistingNoRollback => {
            destination.directory().replace_discarding_destination(
                owner.output.attempt.name(),
                &owner.output.file,
                destination.main(),
            )?
        }
    }
    owner.rename_succeeded = true;
    crate::fault::crash("publication.after_main_rename");
    Ok(())
}

fn synchronize_main(
    owner: &MainAttempt,
    checkpoint: &mut impl FnMut(Point) -> Result<(), Error>,
) -> Result<(), Error> {
    sync_file(&owner.output.file).map_err(error::Error::from)?;
    crate::fault::crash("publication.after_main_sync");
    checkpoint(Point::MainSynced)?;
    owner.output.attempt.destination().directory().sync()?;
    crate::fault::crash("publication.after_main_directory_sync");
    checkpoint(Point::DirectorySynced)
}

fn prove_main(
    owner: &mut MainAttempt,
    checkpoint: &mut impl FnMut(Point) -> Result<(), Error>,
) -> Result<(), Error> {
    owner.output.attempt.destination().directory().verify()?;
    owner.output.verify_main().map_err(Error::Output)?;
    owner.desired_proven = true;
    crate::fault::crash("publication.after_main_proof");
    checkpoint(Point::DesiredProven)?;
    owner
        .reservation
        .verify_after_main(&owner.output)
        .map_err(Error::Reservation)
}

impl PublishedMain {
    // Retirement failure preserves the already-proven main and exact cleanup state.
    #[allow(clippy::result_large_err)]
    pub(crate) fn retire(self) -> Result<PublishedOutput, Failure<RetiringMain>> {
        self.retire_with(|_| Ok(()))
    }

    #[allow(clippy::result_large_err)]
    fn retire_with(
        self,
        mut checkpoint: impl FnMut(Point) -> Result<(), Error>,
    ) -> Result<PublishedOutput, Failure<RetiringMain>> {
        let mut owner = RetiringMain {
            published: self,
            previous_unlinked: false,
            previous_retired_proven: false,
            reservation_unlinked: false,
            directory_synced: false,
            reservation_retired_proven: false,
            housekeeping: Housekeeping::None,
            visible_housekeeping: Vec::new(),
        };
        match retire_steps(&mut owner, &mut checkpoint) {
            Ok(()) => {
                let RetiringMain {
                    published,
                    housekeeping,
                    visible_housekeeping,
                    ..
                } = owner;
                Ok(PublishedOutput {
                    reservation_identity: published.reservation.identity,
                    reservation_header: published.reservation.header,
                    output: published.output,
                    housekeeping,
                    visible_housekeeping,
                })
            }
            Err(cause) => Err(Failure { owner, cause }),
        }
    }
}

fn retire_steps(
    owner: &mut RetiringMain,
    checkpoint: &mut impl FnMut(Point) -> Result<(), Error>,
) -> Result<(), Error> {
    verify_published(&owner.published)?;
    if unlink_previous(owner)? {
        checkpoint(Point::PreviousUnlinked)?;
    }
    unlink_reservation(owner)?;
    checkpoint(Point::ReservationUnlinked)?;
    sync_retirement(owner)?;
    checkpoint(Point::RetirementSynced)?;
    verify_retired(owner)
}

#[cfg(unix)]
fn unlink_previous(owner: &mut RetiringMain) -> Result<bool, Error> {
    let published = &owner.published;
    let Some(previous) = &published.output.previous else {
        owner.previous_retired_proven = true;
        return Ok(false);
    };
    let destination = published.output.attempt.destination();
    previous.verify_private_or_retired(destination, published.output.attempt.name())?;
    if regular_link_count(&previous.file)? == 0 {
        owner.previous_unlinked = true;
        return Ok(false);
    }
    if !destination
        .directory()
        .unlink_exact(published.output.attempt.name(), previous.identity)?
    {
        return Err(NamespaceError::Missing.into());
    }
    owner.previous_unlinked = true;
    let links = regular_link_count(&previous.file)?;
    if links != 0 {
        return Err(Error::PreviousLinkCount(links));
    }
    crate::fault::crash("publication.after_previous_unlink");
    Ok(true)
}

#[cfg(windows)]
fn unlink_previous(owner: &mut RetiringMain) -> Result<bool, Error> {
    use super::gc::{self, Authority};
    use super::gc_codec::Payload;
    use super::{ArtifactKind, CreationSecurity, DirectoryRole};

    let published = &owner.published;
    let Some(previous) = &published.output.previous else {
        owner.previous_retired_proven = true;
        return Ok(false);
    };
    let destination = published.output.attempt.destination();
    previous.verify_private_or_retired(destination, published.output.attempt.name())?;
    let retirement = gc::retire(
        destination.directory(),
        Authority {
            attempt_id: published.output.attempt.attempt_id(),
            ordinal: 0,
            kind: ArtifactKind::PrivateOutput,
            directory_role: DirectoryRole::Destination,
            source_name: published.output.attempt.name(),
            source_file: &previous.file,
            identity: previous.identity,
            creation_security: CreationSecurity {
                kind: super::namespace::CREATION_SECURITY_KIND,
                commitment: destination.security_commitment(),
            },
            payload: Some(Payload {
                byte_length: previous.byte_length,
                sha512: previous.sha512,
                database_id: [0; 16],
                transaction_id: 0,
                commit_nonce: [0; 16],
            }),
        },
    );
    absorb_gc(owner, retirement)?;
    owner.previous_unlinked = true;
    owner.previous_retired_proven = true;
    crate::fault::crash("publication.after_previous_unlink");
    Ok(true)
}

fn verify_published(published: &PublishedMain) -> Result<(), Error> {
    published.output.verify_main().map_err(Error::Output)?;
    published
        .reservation
        .verify_after_main(&published.output)
        .map_err(Error::Reservation)
}

#[cfg(unix)]
fn unlink_reservation(owner: &mut RetiringMain) -> Result<(), Error> {
    let published = &owner.published;
    let destination = published.output.attempt.destination();
    if !destination
        .directory()
        .unlink_exact(destination.coordination(), published.reservation.identity)?
    {
        return Err(NamespaceError::Missing.into());
    }
    owner.reservation_unlinked = true;
    let links = regular_link_count(&published.reservation.file)?;
    if links != 0 {
        return Err(Error::ReservationLinkCount(links));
    }
    crate::fault::crash("publication.after_reservation_unlink");
    Ok(())
}

#[cfg(windows)]
fn unlink_reservation(owner: &mut RetiringMain) -> Result<(), Error> {
    use super::gc::{self, Authority};
    use super::{ArtifactKind, CreationSecurity, DirectoryRole};

    let published = &owner.published;
    let destination = published.output.attempt.destination();
    let retirement = gc::retire(
        destination.directory(),
        Authority {
            attempt_id: published.output.attempt.attempt_id(),
            ordinal: 1,
            kind: ArtifactKind::OwnedCoordination,
            directory_role: DirectoryRole::Destination,
            source_name: destination.coordination(),
            source_file: &published.reservation.file,
            identity: published.reservation.identity,
            creation_security: CreationSecurity {
                kind: super::namespace::CREATION_SECURITY_KIND,
                commitment: destination.security_commitment(),
            },
            payload: None,
        },
    );
    absorb_gc(owner, retirement)?;
    owner.reservation_unlinked = true;
    owner.reservation_retired_proven = true;
    crate::fault::crash("publication.after_reservation_unlink");
    Ok(())
}

fn sync_retirement(owner: &mut RetiringMain) -> Result<(), Error> {
    owner
        .published
        .output
        .attempt
        .destination()
        .directory()
        .sync()?;
    owner.directory_synced = true;
    crate::fault::crash("publication.after_retirement_sync");
    Ok(())
}

fn verify_retired(owner: &mut RetiringMain) -> Result<(), Error> {
    let output = &owner.published.output;
    let destination = output.attempt.destination();
    destination.directory().verify()?;
    destination
        .directory()
        .require_absent(destination.coordination())?;
    if !owner.previous_retired_proven {
        if let Some(previous) = &output.previous {
            previous.verify_retired(destination, output.attempt.name())?;
        }
    }
    owner.previous_retired_proven = true;
    owner.reservation_retired_proven = true;
    output.verify_main().map_err(Error::Output)
}

#[cfg(windows)]
fn absorb_gc(owner: &mut RetiringMain, retirement: super::gc::Retirement) -> Result<(), Error> {
    owner.housekeeping = merge_housekeeping(owner.housekeeping, retirement.housekeeping);
    owner.visible_housekeeping.extend(retirement.visible);
    match retirement.problem {
        Some(problem) => Err(Error::Gc(problem)),
        None => Ok(()),
    }
}

#[cfg(windows)]
fn merge_housekeeping(left: Housekeeping, right: Housekeeping) -> Housekeeping {
    if matches!(left, Housekeeping::Visible) || matches!(right, Housekeeping::Visible) {
        Housekeeping::Visible
    } else if matches!(left, Housekeeping::CrashReappearancePossible)
        || matches!(right, Housekeeping::CrashReappearancePossible)
    {
        Housekeeping::CrashReappearancePossible
    } else {
        Housekeeping::None
    }
}

impl From<NamespaceError> for Error {
    fn from(value: NamespaceError) -> Self {
        Self::Namespace(value)
    }
}

impl From<error::Error> for Error {
    fn from(value: error::Error) -> Self {
        Self::Sdk(value)
    }
}

#[cfg(all(test, unix))]
#[path = "main_file_tests.rs"]
mod tests;

File: v4/go/internal/publication/main_file.go
// Atomic main-name publication and exact reservation retirement (Rust
// publication/main_file.rs). The main file appears under its final
// name with one atomic namespace operation per policy, the file and
// directory synced and re-proved; retirement then retires the replaced
// previous and the coordination reservation and proves the namespace
// empty of them before the result facts are built. The posix arms
// unlink exactly; the windows arms run the authenticated GC
// transition (main_file_windows.go).

package publication

import (
	"github.com/firehol/iprange/v4/go/internal/fault"
	"github.com/firehol/iprange/v4/go/internal/live"
)

// mainPoint is one main-file machine checkpoint (Rust main_file::Point).
type mainPoint uint8

const (
	mainPointMainRenamed mainPoint = iota
	mainPointMainSynced
	mainPointDirectorySynced
	mainPointDesiredProven
	mainPointPreviousUnlinked
	mainPointReservationUnlinked
	mainPointRetirementSynced
)

// mainFailure is one main-file failure carrying the still-owned
// attempt (Rust Failure<MainAttempt>). The owner rides by value so
// the success path of the machine stays on the stack (Rust moves the
// owner; Go copies it into the failure only on the failure path).
type mainFailure struct {
	owner mainAttempt
	cause error
}

// mainAttempt is the publish owner: the output and armed reservation
// plus the exact physical-state flags the recovery decision needs
// (Rust MainAttempt).
type mainAttempt struct {
	output          *preparedOutput
	reservation     armedReservation
	mainCallStarted bool
	renameSucceeded bool
	desiredProven   bool
}

// publishedMain is the proven published output and its reservation
// (Rust PublishedMain).
type publishedMain struct {
	output      *preparedOutput
	reservation armedReservation
}

// retiringMain is the retirement owner: the published main plus the
// exact retirement-progress flags, housekeeping evidence, and the
// optional visible housekeeping artifacts (Rust RetiringMain).
type retiringMain struct {
	published                publishedMain
	previousUnlinked         bool
	previousRetiredProven    bool
	reservationUnlinked      bool
	directorySynced          bool
	reservationRetiredProven bool
	housekeeping             Housekeeping
	visibleHousekeeping      []HousekeepingArtifact
}

// retiringMainFailure is one retirement failure carrying the still-
// owned retiring main (Rust Failure<RetiringMain>).
type retiringMainFailure struct {
	owner retiringMain
	cause error
}

// publishedOutput is the retired publication output (Rust
// PublishedOutput): the main inode and its lifetime lock stay owned
// through result construction, with the exact housekeeping evidence.
type publishedOutput struct {
	outputGuard         *preparedOutput
	reservationIdentity live.FileIdentity
	housekeeping        Housekeeping
	visibleHousekeeping []HousekeepingArtifact
}

// publishProved runs the main-name publication without checkpoints
// (Rust main_file::publish).
func publishProved(output *preparedOutput, reservation armedReservation) (publishedMain, *mainFailure) {
	return publishWith(output, reservation, func(mainPoint) error { return nil })
}

// publishObserved runs the main-name publication with one exact
// checkpoint per physical step (Rust main_file::publish_observed; the
// attempt machine wires the DesiredProven observation here).
func publishObserved(output *preparedOutput, reservation armedReservation, checkpoint func(mainPoint) error) (publishedMain, *mainFailure) {
	return publishWith(output, reservation, checkpoint)
}

func publishWith(output *preparedOutput, reservation armedReservation, checkpoint func(mainPoint) error) (publishedMain, *mainFailure) {
	owner := mainAttempt{
		output:      output,
		reservation: reservation,
	}
	if err := publishSteps(&owner, checkpoint); err != nil {
		return publishedMain{}, &mainFailure{owner: owner, cause: err}
	}
	return publishedMain{output: owner.output, reservation: owner.reservation}, nil
}

func publishSteps(owner *mainAttempt, checkpoint func(mainPoint) error) error {
	if err := verifyBeforeMain(owner); err != nil {
		return err
	}
	if err := renameMain(owner); err != nil {
		return err
	}
	if err := checkpoint(mainPointMainRenamed); err != nil {
		return checkpointFailure(err)
	}
	if err := synchronizeMain(owner, checkpoint); err != nil {
		return err
	}
	return proveMain(owner, checkpoint)
}

// verifyBeforeMain proves the reservation is armed for this output
// and the destination main slot is either the verified previous or
// absent (Rust verify_before_main).
func verifyBeforeMain(owner *mainAttempt) error {
	if err := owner.reservation.verifyBeforeMain(owner.output); err != nil {
		return err
	}
	if owner.output.previous != nil {
		return owner.output.verifyDestinationBeforeMain()
	}
	destination := owner.output.attempt.destinationOf()
	return destination.directory().RequireAbsent(destination.mainName())
}

// renameMain places the private output under the final main name with
// the policy's atomic operation (Rust rename_main).
func renameMain(owner *mainAttempt) error {
	destination := owner.output.attempt.destinationOf()
	owner.mainCallStarted = true
	switch owner.output.policy {
	case reservationPolicyFailIfExists:
		if err := destination.directory().RenameNoReplace(owner.output.attempt.nameOf(), owner.output.file, destination.mainName()); err != nil {
			return err
		}
	case reservationPolicyReplaceExisting:
		if err := destination.directory().RenameExchange(owner.output.attempt.nameOf(), destination.mainName()); err != nil {
			return err
		}
	default: // reservationPolicyReplaceExistingNoRollback
		if err := destination.directory().RenamePlain(owner.output.attempt.nameOf(), destination.mainName()); err != nil {
			return err
		}
	}
	owner.renameSucceeded = true
	fault.Crash("publication.after_main_rename")
	return nil
}

// synchronizeMain durably flushes the main file and the directory at
// the exact Rust steps (Rust synchronize_main).
func synchronizeMain(owner *mainAttempt, checkpoint func(mainPoint) error) error {
	if err := live.SyncFile(owner.output.file); err != nil {
		return err
	}
	fault.Crash("publication.after_main_sync")
	if err := checkpoint(mainPointMainSynced); err != nil {
		return checkpointFailure(err)
	}
	if err := owner.output.attempt.destinationOf().directory().Sync(); err != nil {
		return err
	}
	fault.Crash("publication.after_main_directory_sync")
	return checkpoint(mainPointDirectorySynced)
}

// proveMain re-proves the directory and the main output, remembers the
// desired proof, and re-proves the reservation (Rust prove_main).
func proveMain(owner *mainAttempt, checkpoint func(mainPoint) error) error {
	if err := owner.output.attempt.destinationOf().directory().Verify(); err != nil {
		return err
	}
	if err := owner.output.verifyMain(); err != nil {
		return err
	}
	owner.desiredProven = true
	fault.Crash("publication.after_main_proof")
	if err := checkpoint(mainPointDesiredProven); err != nil {
		return checkpointFailure(err)
	}
	return owner.reservation.verifyAfterMain(owner.output)
}

// retire retires one published main without checkpoints (Rust
// PublishedMain::retire). retire_observed shares the same arms; the
// windows gc wiring observes housekeeping through the GC transition.
func (p *publishedMain) retire() (publishedOutput, *retiringMainFailure) {
	return retireMain(*p, func(mainPoint) error { return nil })
}

func retireMain(p publishedMain, checkpoint func(mainPoint) error) (publishedOutput, *retiringMainFailure) {
	owner := retiringMain{published: p}
	if err := retireSteps(&owner, checkpoint); err != nil {
		return publishedOutput{}, &retiringMainFailure{owner: owner, cause: err}
	}
	return publishedOutput{
		outputGuard:         owner.published.output,
		reservationIdentity: owner.published.reservation.identity,
		housekeeping:        owner.housekeeping,
		visibleHousekeeping: owner.visibleHousekeeping,
	}, nil
}

func retireSteps(owner *retiringMain, checkpoint func(mainPoint) error) error {
	if err := verifyPublished(owner.published); err != nil {
		return err
	}
	unlinked, err := unlinkPrevious(owner)
	if err != nil {
		return err
	}
	if unlinked {
		if err := checkpoint(mainPointPreviousUnlinked); err != nil {
			return checkpointFailure(err)
		}
	}
	if err := unlinkReservation(owner); err != nil {
		return err
	}
	if err := checkpoint(mainPointReservationUnlinked); err != nil {
		return checkpointFailure(err)
	}
	if err := syncRetirement(owner); err != nil {
		return err
	}
	if err := checkpoint(mainPointRetirementSynced); err != nil {
		return checkpointFailure(err)
	}
	return verifyRetired(owner)
}

// verifyPublished re-proves the published main and its reservation
// before any retirement syscall (Rust verify_published).
func verifyPublished(published publishedMain) error {
	if err := published.output.verifyMain(); err != nil {
		return err
	}
	return published.reservation.verifyAfterMain(published.output)
}

// syncRetirement synchronizes the directory after the unlinks (Rust
// sync_retirement).
func syncRetirement(owner *retiringMain) error {
	if err := owner.published.output.attempt.destinationOf().directory().Sync(); err != nil {
		return err
	}
	owner.directorySynced = true
	fault.Crash("publication.after_retirement_sync")
	return nil
}

// verifyRetired proves the directory, the absent coordination name,
// the retired previous, and the still-intact main (Rust
// verify_retired).
func verifyRetired(owner *retiringMain) error {
	output := owner.published.output
	destination := output.attempt.destinationOf()
	if err := destination.directory().Verify(); err != nil {
		return err
	}
	if err := destination.directory().RequireAbsent(destination.coordinationName()); err != nil {
		return err
	}
	if !owner.previousRetiredProven {
		if previous := output.previous; previous != nil {
			if err := previous.verifyRetired(destination, output.attempt.nameOf()); err != nil {
				return err
			}
		}
	}
	owner.previousRetiredProven = true
	owner.reservationRetiredProven = true
	return output.verifyMain()
}
